dataloader/dataloader_impl.hpp

```diff
@@ -0,0 +1,151 @@
+/**
+ * @file dataloader_impl.hpp
+ * @author Kartik Dutt
+ * 
+ * Implementation of DataLoader.
+ * 
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MODELS_DATALOADER_IMPL_HPP
+#define MODELS_DATALOADER_IMPL_HPP
+
+#include "dataloader.hpp"
+
+using namespace mlpack;
+
+template<
+  typename DatasetX,
+  typename DatasetY,
+  class ScalerType
+>DataLoader<
+    DatasetX, DatasetY, ScalerType
+>::DataLoader()
+{
+  // Nothing to do here.
+}
+
+template<
+  typename DatasetX,
+  typename DatasetY,
+  class ScalerType
+>DataLoader<
+    DatasetX, DatasetY, ScalerType
+>::DataLoader(const std::string& dataset,
+              const bool shuffle,
+              const double ratio,
+              const bool useScaler,
+              const std::vector<std::string> augmentation,
+              const double augmentationProbability)
+{
+  InitializeDatasets();
+  if (datasetMap.count(dataset))
+  {
+    // Use utility functions to download the dataset.
+    DownloadDataset(dataset);
+
+    if (datasetMap[dataset].loadCSV)
+    {
+      LoadCSV(datasetMap[dataset].trainPath, true, shuffle, ratio, useScaler,
+              datasetMap[dataset].dropHeader,
+              datasetMap[dataset].startTrainingInputFeatures,
+              datasetMap[dataset].endTrainingInputFeatures,
+              datasetMap[dataset].endTrainingPredictionFeatures,
+              datasetMap[dataset].endTrainingPredictionFeatures);
+
+      if (dataset == "mnist")
```

Comment:
We might want to apply a policy based design here, like a dataset specific class to apply some specific functions after the data is loaded, something like:

```c++
class Mnist {
  static void Update(...)
  {

  }
}
```

let me know what you think

Reply:
Yeah, we could do that. Just one more thing, I am currently storing a class that determines type of dataset and if it is of `csv` format we store rows to drop, rows to use for prediction etc. I think we should keep that as it is or would you suggest having them in the class you mentioned above.

Reply:
This would definitely more sense when more datasets are added.

Reply:
Hey @zoq, I was thinking if we could do something like the following :

```
class Preprocess
{
   public:
   static void MNIST(input-args)
   {
     // Preprocess dataset.
    }
}
```

That way we don't have to have a separate class for each dataset we add here, and eventually save the lines that would have to be maintained.

Kindly let me know if this makes sense. Thanks a lot.

Regards,
Kartik.

Reply:
Sure, that looks reasonable to me.

Reply:
Great, Thanks a lot.